### PR TITLE
Add additional options for automargin property

### PR DESCRIPTION
**Pull request by nickmelnikov82**
**Feature**: Add new options "width" and "height" for the automargin property, which would limit the automargin calculation to one dimension. This way we could make sure the labels are never cut off and that there isn't any whitespace on the left or right border of the figure.

### CODE REVIEW

File: src/plots/cartesian/axes.js [hunk 1]
```diff
@@ -2636,6 +2638,35 @@ axes.drawOne = function(gd, ax, opts) {
...
+    Object.keys(mapping).forEach(function(key) {
+        if(automargin.includes(key)) {
+            mapping[key].forEach(function(item) {
+                keepMargin.push(item);
+            });
+        }
+    });
+
+    Object.keys(push).forEach(function(key) {
+        if(key.length !== 2 && !keepMargin.includes(key)) {
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js:2663]: What's the `length !== 2` about?

Comment (alexcjohnson) [reply src/plots/cartesian/axes.js:2663]: OIC there's `xl`, `xr`, `yt`, `yb`. But don't we want to always include `x` and `y` as well? I've forgotten exactly what those do, and doesn't look like we ever documented this... but it does appear that `x` and `y` are the fallbacks when for whatever reason we didn't provide the more specific 2-letter variants

https://github.com/plotly/plotly.js/blob/b194d8c3ef503ac1989666547cccb65dc60fb44f/src/plots/plots.js#L1931-L1934

File: src/plots/cartesian/layout_attributes.js [hunk 2]
```diff
@@ -625,7 +625,14 @@ module.exports = {
...
     automargin: {
-        valType: 'boolean',
+        valType: 'enumerated',
+        values: [
+            true, false, 'height', 'width',
+            'top', 'bottom', 'left', 'right',
+            'top left', 'top width', 'top right',
+            'left height', 'right height',
+            'bottom left', 'bottom width', 'bottom right',
+        ],
```
Comment (alexcjohnson) [inline src/plots/cartesian/layout_attributes.js:635]: Let's use a flaglist:
```js
valType: 'flaglist',
flags: ['height', 'width', 'left', 'right', 'top', 'bottom'],
extras: [true, false]
```
That way the order doesn't matter (you use `+` - `bottom+left` and `left+bottom` are both valid), and users could even do `bottom+left+right` which would mean the same thing as `bottom+width`

File: src/plots/cartesian/axes.js [hunk 3]
```diff
@@ -2636,6 +2638,35 @@ axes.drawOne = function(gd, ax, opts) {
...
+        width: ['x', 'r', 'l'],
+        height: ['y', 't', 'b'],
+        right: ['r'],
+        left: ['l'],
+        top: ['t'],
+        bottom: ['b']
+    };
+
+    Object.keys(mapping).forEach(function(key) {
+        if(automargin.includes(key)) {
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js:2655]: 😢 still need to support IE11 so [can't use .includes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes#browser_compatibility)

File: src/plots/cartesian/layout_attributes.js (hunk 2 above)
Comment (nicolaskruchten) [reply src/plots/cartesian/layout_attributes.js:635]: Are there pre-existing instances of "flaglist or boolean" in the schema? I want to make sure the Python codegen will not choke on this.

File: src/plots/cartesian/axes.js (hunk 3 above)
Comment (nickmelnikov82) [reply src/plots/cartesian/axes.js]: fixed

File: src/plots/cartesian/layout_attributes.js (hunk 2 above)
Comment (nickmelnikov82) [reply src/plots/cartesian/layout_attributes.js:635]: flaglist does not support boolean parameters https://github.com/plotly/plotly.js/blob/b194d8c3ef503ac1989666547cccb65dc60fb44f/src/lib/coerce.js#L218

File: src/plots/cartesian/axes.js (hunk 1 above)
Comment (nickmelnikov82) [reply src/plots/cartesian/axes.js]: Yes, it is for these properties. Corrected the code, made the removal of the property if it  is not in the keep list.

File: src/plots/cartesian/layout_attributes.js (hunk 2 above)
Comment (alexcjohnson) [reply src/plots/cartesian/layout_attributes.js:635]: OK that's pretty good evidence that we don't currently have any flaglists with non-string extras 😉
We could allow this in `coerce` pretty easily by pushing the [extras test](https://github.com/plotly/plotly.js/blob/b194d8c3ef503ac1989666547cccb65dc60fb44f/src/lib/coerce.js#L222-L225) up above that non-string short-circuit.

@nicolaskruchten your concern is warranted, plotly.py also includes the [string requirement up front](https://github.com/plotly/plotly.py/blob/0279560d38381ed45afb79cb2aaf567b3f4c82a4/packages/python/plotly/_plotly_utils/basevalidators.py#L1838-L1839). But I really do think this attribute would be better as a flaglist, I'll be happy to modify the Python validator to allow non-string extras.

I suppose the alternative would be to make this a new attribute like `automargindirections` and leave `automargin` as a boolean, but that seems a bit cumbersome...

Comment (nicolaskruchten) [reply src/plots/cartesian/layout_attributes.js:635]: I'm fine with tweaking Plotly.py for this 👍

Comment (nickmelnikov82) [reply src/plots/cartesian/layout_attributes.js]: Changed the schema.

**Review by alexcjohnson — APPROVED** [review]
💃 This looks great to me. It'll just need a draftlog entry, @archmoj anything from your side?

I'll make the required plotly.py change for non-string flaglist extras myself.

File: src/lib/coerce.js [hunk 4]
```diff
@@ -215,14 +215,14 @@ exports.valObjectMeta = {
...
-                return;
-            }
             if((opts.extras || []).indexOf(v) !== -1) {
                 propOut.set(v);
                 return;
             }
+            if(typeof v !== 'string') {
+                propOut.set(dflt);
+                return;
+            }
```
Comment (archmoj) [inline src/lib/coerce.js:225]: Could you please elaborate why this change is needed?

Comment (alexcjohnson) [reply src/lib/coerce.js:225]: Just reversing the order of short-circuits to support non-string extras https://github.com/plotly/plotly.js/pull/6193#discussion_r879534123

Comment (nickmelnikov82) [reply src/lib/coerce.js:225]: As @alexcjohnson wrote:
```
OK that's pretty good evidence that we don't currently have any flaglists with non-string extras 😉
We could allow this in coerce pretty easily by pushing the extras test up above that non-string short-circuit.
```

Without this, the flaglist coerce function does not pass non-string values.

File: src/plots/cartesian/axes.js [hunk 5]
```diff
@@ -2636,6 +2638,39 @@ axes.drawOne = function(gd, ax, opts) {
     return Lib.syncOrAsync(seq);
 };
 
+function keepSelectedAutoMargin(automargin, push) {
+    if(typeof automargin === 'boolean') return push;
+
+    var keepMargin = [];
+    var mapping = {
```
Comment (archmoj) [inline src/plots/cartesian/axes.js:2645]: It looks like `mapping` is a constant. 
Let's use all caps name and move it outside the scope of this function to avoid creation of the object every time calling the function. 
Thanks.

File: draftlogs/6193_add.md [hunk 6]
```diff
@@ -0,0 +1 @@
+ - Add new options ('width', 'height', 'left', 'right', 'top', 'bottom') for the automargin property [[#6193](https://github.com/plotly/plotly.js/pull/6193)].
```
Comment (archmoj) [inline draftlogs/6193_add.md:1]:
```suggestion
 - Add flaglist options including "left", "right", "top", "bottom", "width" and "height" to control the direction of `automargin` on cartesian axes [[#6193](https://github.com/plotly/plotly.js/pull/6193)]
 ```

File: src/plots/cartesian/axes.js (hunk 5 above)
Comment (nickmelnikov82) [reply src/plots/cartesian/axes.js]: fixed

File: src/plots/cartesian/axes.js [hunk 7]
```diff
@@ -2622,6 +2622,8 @@ axes.drawOne = function(gd, ax, opts) {
             rangeSliderPush = Registry.getComponentMethod('rangeslider', 'autoMarginOpts')(gd, ax);
         }
 
+        keepSelectedAutoMargin(ax.automargin, push);
```
[the remark this answers is not in the document]
Comment (nickmelnikov82) [reply src/plots/cartesian/axes.js]: I don't see the point of using a mirror if the result would still be cut off. So there is no need to filter the `mirrorPush`.

File: src/plots/cartesian/layout_attributes.js (hunk 2 above)
Comment (archmoj) [reply src/plots/cartesian/layout_attributes.js]: > OK that's pretty good evidence that we don't currently have any flaglists with non-string extras wink We could allow this in `coerce` pretty easily by pushing the [extras test](https://github.com/plotly/plotly.js/blob/b194d8c3ef503ac1989666547cccb65dc60fb44f/src/lib/coerce.js#L222-L225) up above that non-string short-circuit.
> 
> @nicolaskruchten your concern is warranted, plotly.py also includes the [string requirement up front](https://github.com/plotly/plotly.py/blob/0279560d38381ed45afb79cb2aaf567b3f4c82a4/packages/python/plotly/_plotly_utils/basevalidators.py#L1838-L1839). But I really do think this attribute would be better as a flaglist, I'll be happy to modify the Python validator to allow non-string extras.
> 
> I suppose the alternative would be to make this a new attribute like `automargindirections` and leave `automargin` as a boolean, but that seems a bit cumbersome...

Having a separate attribute could be safer for Python, R, Scala, etc. APIs.
It would be nice to possibly test this at `plotly.py` level before merging this PR.

File: src/plots/cartesian/axes.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexcjohnson) [reply src/plots/cartesian/axes.js]: Actually I think it's correct to filter `mirrorPush` as well. The whole `mirrorPush` construction is inside the [`if(ax.automargin)` block](https://github.com/plotly/plotly.js/blob/cf2623792345ed7bd330ee9016ef2aefbf76fd14/src/plots/cartesian/axes.js#L2561) - implying that if you disable automargin, we're not going to push the margins for mirror axes regardless of whether they end up cut off. Therefore you should be able to choose which directions this applies to just as you can the regular axis automargin.

File: src/plots/cartesian/axes.js [hunk 8]
```diff
@@ -2636,6 +2649,31 @@ axes.drawOne = function(gd, ax, opts) {
...
 };
 
+function filterPush(push, automargin) {
+    if(!push) return push;
+
+    var keepMargin = [];
+
+    Object.keys(MARGIN_MAPPING).forEach(function(key) {
+        if(automargin.indexOf(key) !== -1) {
+            MARGIN_MAPPING[key].forEach(function(item) {
```
Comment (archmoj) [inline src/plots/cartesian/axes.js:2659]: Can't we simply use one [`concact`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat) here instead of the inner loop and push?

File: src/plots/cartesian/axes.js [hunk 9]
```diff
@@ -2622,6 +2630,11 @@ axes.drawOne = function(gd, ax, opts) {
             rangeSliderPush = Registry.getComponentMethod('rangeslider', 'autoMarginOpts')(gd, ax);
         }
 
+        if(typeof ax.automargin === 'string') {
+            filterPush(push, ax.automargin);
+            filterPush(mirrorPush, ax.automargin);
```
Comment (archmoj) [inline src/plots/cartesian/axes.js:2635]: How about `rangeSliderPush`? Should we filter that one too?

Comment (alexcjohnson) [reply src/plots/cartesian/axes.js:2635]: ha @nickmelnikov82 asked me the same question, my answer:

> that one isn’t controlled by `ax.automargin` so I think no

There are others that aren't even in this section of code, like legends and colorbars. `ax.automargin` is only about "are the axes themselves allowed to increase the margins."

**Comment by alexcjohnson** [comment]
Created https://github.com/plotly/plotly.py/pull/3749 to allow non-string flaglist extras on the Python side

File: src/plots/cartesian/axes.js (hunk 9 above)
Comment (alexcjohnson) [reply src/plots/cartesian/axes.js:2635]: Oh that's interesting... in principle I like this, one place to control all of this rather than requiring you to set it in each axis separately. But it opens a can of worms: by default axes have `automargin: false` while colorbars and legends (and rangesliders) by default DO increase the margins, and putting this at the top level would imply that it applies to everything. So it would need a default value like `'legacy'` or something that gets inherited differently by axes and other objects, and then we would need to extend support for disabling automargin to all these other objects as well.

All that to say: it's a good idea, but it's going to be a bunch of work, so let's not do it now. What we're doing in this PR would be fully compatible with adding that sometime in the future.

File: src/plots/cartesian/axes.js [hunk 10]
```diff
@@ -2636,6 +2649,25 @@ axes.drawOne = function(gd, ax, opts) {
...
+        return data;
+    }, {});
+
+    Object.keys(push).forEach(function(key) {
+        if(!keepMargin[key]) {
+            if(key.length === 1) push[key] = 0;
+            else delete push[key];
+        }
+    });
+    return push;
```
Comment (archmoj) [inline src/plots/cartesian/axes.js:2668]:
```suggestion
```

**Comment by archmoj** [comment]
Thanks very much for the PR :dancer:
